Evict pods w/o rate-limit when cloud says node is gone.

cjcullen opened the pull request:
This decouples scale-down on cloud clusters from rate-limited eviction.

cjcullen committed:
Evict pods w/o rate-limit when cloud says node is gone.

cjcullen commented:
@alex-mohr

smarterclayton commented:
Looks good on observation but want another set of eyes - the locks look right but I don't want to deadlock here.

gmarek commented on pkg/controller/node/nodecontroller.go:
```diff
@@ -243,15 +243,10 @@ func (nc *NodeController) Run(period time.Duration) {
-			remaining, err := nc.deletePods(value.Value)
-			if err != nil {
-				utilruntime.HandleError(fmt.Errorf("unable to evict node %q: %v", value.Value, err))
+			if err := nc.evictPodsNoRateLimit(value.Value); err != nil {
```
The name of the function is misleading - this call is rate limited, as 'Try' is. Please rename it to plain 'evictPods' or something like that (actually we need some better naming for all those functions - having `deletePods`, `terminatePods` and `evictPods` is not very readable, but I'll handle this separately).

cjcullen commented:
I realized that by aiming for a cuter refactor, I just moved the bottleneck, because the terminationEvictor is rate-limited exactly like the podEvictor.

I changed this to:
Try to directly deletePods.
Then try to directly terminatePods.
If deletePods fails, queue the node on the podEvictor.
If terminatePods fails, queue the node on the terminationEvictor.

smarterclayton commented on pkg/controller/node/nodecontroller.go:
```diff
@@ -482,9 +482,33 @@ func (nc *NodeController) monitorNodeStatus() error {
+							if !remaining {
+								return
+							}
+							completed, _, err := nc.terminatePods(nodeName, time.Now())
```
In practice this method will always return completed false.  It would be better to go straight to the graceful check queue than to call terminate pods.

smarterclayton replied:
Because that will release the evictor lock

cjcullen replied:
Done. IIUC, that will immediately sigterm all pods. sigkill will still be rate-limited, but unless the pods are stubborn, they should be gone by the next time through this loop anyways. Sound right?

gmarek commented:
@cjcullen - please apply @smarterclayton comment.

cjcullen commented:
@gmarek @smarterclayton: Changes applied. Would like to get this in for 1.2 to help with scaling down large clusters.

gmarek commented on pkg/controller/node/nodecontroller.go:
```diff
@@ -483,9 +483,24 @@ func (nc *NodeController) monitorNodeStatus() error {
+								return
+							}
+							// Queue for termination.
+							nc.terminationEvictor.Add(nodeName)
```
I don't think we should rate limit it at all. IIUC this if branch is responsible for deleting stuff from etcd after node is gone. Hence no sigterms or sigkills will be emitted. You can reuse deletePods/terminatePods functions, but you don't need to rate limit any of them.

cjcullen replied:
Ok. Switched it to immediately attempt termination (falling back on queuing in case of an error).

cjcullen commented:
@k8s-bot e2e test this issue: #21995

smarterclayton commented on pkg/controller/node/nodecontroller.go:
```diff
@@ -480,9 +480,26 @@ func (nc *NodeController) monitorNodeStatus() error {
+							if !remaining {
+								return
+							}
+							// Immediately terminate pods.
```
I was just suggesting adding to terminationEvictor here, instead of invoking `terminatePods`.

cjcullen replied:
That's what I had this morning, but @gmarek convinced me otherwise. We don't expect any *actual* pods to be running, though they may still be represented in the apiserver. If I understand the system, it should come out fine whether we queue or directly terminate. 

If we try to directly terminate and it fails for whatever reason, we'll bail, but next scan through the nodes we should attempt to evict/terminate again.

If we add to the evictor queue, we might get more retries before the next full node scan, but we may also be backing up the rate-limited piece for no good reason.

If that description doesn't match with what you think would actually happen, I'm happy to change the PR.

gmarek replied:
Yup - I don't think that rate limiting makes any sense here. This is a branch responsible for evicting Pods from Nodes that are gone from the cloud provider's perspective. If for whatever reason we would be able to contact this Node it would mean that there's some serious bug in the control plane of the given provider.

alex-mohr commented:
@k8s-bot unit test this

alex-mohr commented:
FYI, filed #22291 for the "no data" from Jenkins unit/integration that caused the re-test request.

cjcullen commented:
Reapplying the LGTM (nothing new here since @gmarek LGTMed it).

cjcullen commented:
Pulling LGTM. This didn't actually fix the problem. All nodes still end up going through the deletion queue.

lavalamp commented:
@eparis we need to make merge-bot re-check LGTM status right before merging :/

@cjcullen please roll back if you didn't actually want this merged!

lavalamp commented:
Why is it important to not rate limit these deletions?

cjcullen commented:
If kubelet has stopped reporting status, and the cloudprovider is claiming the node VM no longer exists, we shouldn't need to rate limit pod eviction.

Our current scale-down method in GCE is "change the size of the MIG." Right now, that scale down is limited to `--deleting-pods-qps` (defaulted to 0.1 [here](https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-controller-manager/app/options/options.go#L124)). This means that for Nodes that we are ~100% sure are intentionally deleted, we will only remove them at 6/minute.

This really only affects Very Large Clusters, but that is a supported use case in 1.2.

gmarek commented:
@cjcullen - what's the problem with this PR?

cjcullen commented:
@gmarek: When we actually tested it on a 1000 node cluster, the nodes still ended up being rate-limited. `deletePods` would return that no pods were remaining, but `hasPods` would still say there were pods left. @alex-mohr for any more details.

gmarek commented:
OK, thanks. If you consider it a blocker for v1.2 I can probably take it.
